Develop

EvgeniyTomilov opened the pull request.

EvgeniyTomilov committed:
тз-11 Checkstyle

EvgeniyTomilov committed:
тз-12 add-search

EvgeniyTomilov committed:
Merge remote-tracking branch 'origin/add-search' into add-search

Slavaniki committed:
add delete endpoint

Slavaniki committed:
delete unused import

Ekaterina-Kirpi committed:
популярные фильмы по жанру и году

AnnaTsepeleva committed:
Push add-reviews

EvgeniyTomilov committed:
тз-12 add-search добавлен перегруженный метод rowMapFilms

Ekaterina-Kirpi committed:
популярные фильмы по жанру и году

Ekaterina-Kirpi committed:
Синтатиксические ошибки

Slavaniki committed:
add throw 404

EvgeniyTomilov committed:
Merge pull request #7 from EvgeniyTomilov/add-remove-endpoint

add delete endpoints

EvgeniyTomilov committed:
Merge branch 'develop' into add-search

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java

Ekaterina-Kirpi committed:
Синтатиксические ошибки

Ekaterina-Kirpi committed:
Merge branch 'develop' into add-most-populars

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java

LipatovKir committed:
добавление режиссера

EvgeniyTomilov committed:
Merge pull request #8 from EvgeniyTomilov/add-reviews

Push add-reviews

Ekaterina-Kirpi committed:
Merge branch 'develop' into add-most-populars

AnnaTsepeleva committed:
Commit by AS

LipatovKir committed:
добавление режиссера/есты проходят

LipatovKir committed:
Merge branch 'develop' into add-director

LipatovKir committed:
Update FilmController.java

LipatovKir committed:
Update FilmController.java

LipatovKir committed:
Поправить schema.sql

LipatovKir committed:
добавление режиссера/исправить недочеты

LipatovKir committed:
Update DirectorDbStorage.java

EvgeniyTomilov committed:
Merge pull request #10 from EvgeniyTomilov/add-director

Add director

Ekaterina-Kirpi committed:
Merge branch 'develop' into add-most-populars

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java
#	src/main/resources/schema.sql

Slavaniki committed:
add recommendations

Ekaterina-Kirpi committed:
JOIN films_directors
JOIN directors

Ekaterina-Kirpi committed:
популярные фильмы с директором

LipatovKir committed:
добавление общих фильмов

LipatovKir committed:
добавление общих фильмов/тесты проходят

Ekaterina-Kirpi committed:
популярные фильмы с директором

EvgeniyTomilov committed:
Merge branch 'develop' into add-search

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java
#	src/main/java/ru/yandex/practicum/filmorate/model/Film.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java

EvgeniyTomilov committed:
Merge pull request #13 from EvgeniyTomilov/add-recommendations

add recommendations

EvgeniyTomilov committed:
Merge branch 'develop' into add-search

EvgeniyTomilov committed:
Merge pull request #15 from EvgeniyTomilov/add-common-films

Add common films

Ekaterina-Kirpi committed:
feed

Ekaterina-Kirpi committed:
feed

EvgeniyTomilov committed:
тз-12 add-search добавлена диаграмма падает 40 тестов поиск по директору не написан

EvgeniyTomilov committed:
Merge branch 'develop' into add-search

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java

Ekaterina-Kirpi committed:
feed

Ekaterina-Kirpi committed:
Merge branch 'add-most-populars' into develop

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java

Ekaterina-Kirpi committed:
style

Ekaterina-Kirpi committed:
style

EvgeniyTomilov committed:
тз-12 add-search добавлена диаграмма падает 40 тестов поиск по директору не написан

EvgeniyTomilov committed:
Update README.md

Ekaterina-Kirpi committed:
add feed

EvgeniyTomilov committed:
Merge pull request #16 from EvgeniyTomilov/add-most-populars

Add most populars

EvgeniyTomilov committed:
тз-12 add-search добавлена диаграмма падает 40 тестов поиск по директору не написан

EvgeniyTomilov committed:
Merge remote-tracking branch 'origin/add-search' into add-search

EvgeniyTomilov committed:
Merge remote-tracking branch 'origin/develop' into develop

EvgeniyTomilov committed:
Merge branch 'develop' into add-search

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java

EvgeniyTomilov committed:
тз-12 add-search добавлена диаграмма падает 40 тестов поиск по директору не написан

Ekaterina-Kirpi committed:
Merge branch 'develop' into add-feed

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java
#	src/main/java/ru/yandex/practicum/filmorate/service/UserService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/user/InMemoryUserStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/user/UserStorage.java
#	src/main/resources/schema.sql

EvgeniyTomilov committed:
тз-12 add-search добавлена диаграмма падает 40 тестов поиск по директору не написан

EvgeniyTomilov committed:
тз-12 add-search  падает 10 тестов

EvgeniyTomilov committed:
тз-12 add-search  падает 6 тестов

AnnaTsepeleva committed:
Commit by AS

Ekaterina-Kirpi committed:
Merge remote-tracking branch 'origin/add-feed' into add-feed

Ekaterina-Kirpi committed:
add feed

Ekaterina-Kirpi committed:
style

EvgeniyTomilov committed:
тз-12 add-search  падает 4 тестов

EvgeniyTomilov committed:
Merge pull request #17 from EvgeniyTomilov/add-feed

Add feed

EvgeniyTomilov committed:
Merge remote-tracking branch 'origin/develop' into develop

EvgeniyTomilov committed:
Merge branch 'develop' into add-search

# Conflicts:
#	src/main/resources/schema.sql

EvgeniyTomilov committed:
тз-12 add-search  тесты пройдены

EvgeniyTomilov committed:
тз-12 add-search  checkStyle

EvgeniyTomilov committed:
тз-12 add-search  checkStyle

EvgeniyTomilov committed:
Merge pull request #18 from EvgeniyTomilov/add-search

Add search

EvgeniyTomilov committed:
тз-12 восстановил @Qualifier  и  InMemoryFilmStorage,

EvgeniyTomilov committed:
тз-12 восстановил @Qualifier  и  InMemoryFilmStorage,

EvgeniyTomilov committed:
тз-12 восстановил @Qualifier  и  InMemoryFilmStorage,
не проходит 7 тестов

EvgeniyTomilov committed:
тз-12
не проходит 3 тестов

Slavaniki committed:
change delete methods

Slavaniki committed:
Merge remote-tracking branch 'origin/add-search' into add-search

# Conflicts:
#	src/main/resources/schema.sql

EvgeniyTomilov committed:
тз-12
не проходит 3 тестов

EvgeniyTomilov committed:
Merge pull request #19 from EvgeniyTomilov/add-search

Add search

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/DirectorController.java:
```diff
@@ -0,0 +1,43 @@
+
+    @GetMapping
+    public List<Director> getAllDirectors() {
+        return directorService.getAllDirectors();
```
В каждый метод каждого контроллера стоит добавить логирование входящих запросов

Slavaniki replied:
добавлено

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -63,8 +57,37 @@ public void deleteLike(@PathVariable Long id, @PathVariable Long userId) {
 
     @GetMapping("/popular")
-    public List<Film> getListPopularFilms(@RequestParam(value = "count", required = false, defaultValue = "10") Integer count) {
+    public List<Film> getListPopularFilms(@RequestParam(value = "count", required = false, defaultValue = "10") Integer count,
```
При наличии `defaultValue` в `required = false` нет необходимости
Аналогично и для остальных мест

EvgeniyTomilov replied:
лишние required = false удалены.

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -63,8 +57,37 @@ public void deleteLike(@PathVariable Long id, @PathVariable Long userId) {
+
+    @DeleteMapping("/{filmId}")
+    public void deleteFilmById(@PathVariable Long filmId) {
+        log.info("Удаление фильма по id:" + filmId + "...");
```
Думаю, необходимости печатать в лог `"..."` нет особой необходимости

EvgeniyTomilov replied:
Мы командой решили так оставить

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -63,8 +57,37 @@ public void deleteLike(@PathVariable Long id, @PathVariable Long userId) {
+    }
+
+    @GetMapping("/search")
+    public List<Film> searchFilms(@RequestParam @NotBlank String query,
```
Для того, чтобы аннотации валидации в параметрах методов - например `@NotBlank` работали, то необходимо разместить аннотацию `@Validated` над классом

EvgeniyTomilov replied:
исправлено

KirPyt replied:
Не увидел `@Validated`

LipatovKir replied:
исправлено

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -28,13 +32,6 @@ public User updateUser(@Valid @RequestBody User user) {
-
     @GetMapping
     public Collection<User> getAllUsers() {
         log.info("Вызов всех пользователей...");
```
В методе `getUserById`, `addFriend`, `deleteFriend`,  `getListOfFriends`, `getListSharedFriends` нет смысла размещать аннотацию `@Valid`, так как мы передаем обычный `Long` и валидировать там нечего)

AnnaTsepeleva replied:
Убрали.

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -73,4 +70,25 @@ public Collection<User> getListSharedFriends(@Valid @PathVariable Long id, @Vali
+    }
+
+    @GetMapping("/{id}/recommendations")
+    public Collection<Film> getRecommendations(@Valid @PathVariable Long id) {
```
Здесь тоже `@Valid` не нужен)

AnnaTsepeleva replied:
Тоже убрали

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/error/ErrorHandler.java:
```diff
@@ -31,6 +33,13 @@ public Map<String, String> handleNotFoundError(final NoSuchElementException e) {
+
     @ExceptionHandler({IllegalArgumentException.class, NullPointerException.class})
     @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
     public Map<String, String> handleInternalError(final RuntimeException e) {
```
В исключении, где Вы ловите 500-ую ошибку, надо ловить ошибки класса `Throwable`. Ошибка 500 определяет ошибки, вызванные разными причинами на сервере (это проверяемые и непроверяемые исключения), а ошибки `RuntimeException`, которые Вы ловите в данный момент, являются только проверяемыми.
Более подробно Вы можете почитать здесь:
https://www.baeldung.com/java-exceptions
(Раздел Exception Hierarchy)

Так же из `@ExceptionHandler` надо будет убрать перехватываемые классы исключений, когда замените основное исключение на `Throwable`

AnnaTsepeleva replied:
Поправила на Throwable, но если убирать перехватываемые, то он на все 500 ошибку отдает.

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/error/ErrorResponse.java:
```diff
@@ -0,0 +1,20 @@
+
+public class ErrorResponse {
+
+    String error;
```
Все поля надо инкапсулировать

AnnaTsepeleva replied:
исправила

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/Director.java:
```diff
@@ -0,0 +1,21 @@
+public class Director {
+
+    @PositiveOrZero
+    Long id;
```
Надо инкапсулировать поле

AnnaTsepeleva replied:
исправила

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/Event.java:
```diff
@@ -0,0 +1,18 @@
+import lombok.Builder;
+import lombok.Data;
+
+//@FieldDefaults(level = AccessLevel.PRIVATE)
```
Неиспользуемый код надо убрать

EvgeniyTomilov replied:
Неиспользуемый код убран

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/DirectorService.java:
```diff
@@ -0,0 +1,68 @@
+    private final DirectorStorage directorStorage;
+
+    public Director getDirectorById(Long id) {
+        Optional<Director> director = Optional.ofNullable(directorStorage.getDirectorById(id));
```
Отличное решение использовать `Optional `

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/Operations.java:
```diff
@@ -0,0 +1,8 @@
+package ru.yandex.practicum.filmorate.model;
+
+public enum
+Operations {
```
Отличное решение сформировать перечисление под данные типы операций
Аналогично круто, что Вы поступили так со всеми остальными значениями (`EventType`)

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/feed/Operation.java:
```diff
@@ -0,0 +1,7 @@
+package ru.yandex.practicum.filmorate.model.feed;
+
+public enum Operation {
```
Зачем нужно данное перечисление, если уже есть `Operations`?

EvgeniyTomilov replied:
инам удален

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -26,9 +32,14 @@ public class FilmService {
     @Qualifier(value = "userDbStorage")
```
Реализации хранилищ `inMemory` надо удалить и аннотации `@Qualifier` убрать

AnnaTsepeleva replied:
исправила

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -26,9 +32,14 @@ public class FilmService {
     @Qualifier(value = "userDbStorage")
     private UserStorage userStorage;
```
Рекомендую добавить `final`

AnnaTsepeleva replied:
исправила

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -66,44 +78,82 @@ public Collection<Film> getAllFilms() {
 
-    public List<Film> getListPopularFilms(Integer count) {
+    public List<Film> getListPopularFilms(Integer count, Integer genreId, Integer year) {
         return likesStorage.getTopFilmLikes()
```
Надо переписать получение фильмов на 1 запрос к БД
Дальнейший отбор значений и их маппинг через `stream` не нужен

Ekaterina-Kirpi replied:
Если не считать режиссеров для каждого случая, то запрос к БД 1, просто учесть все случаи иначе не выходит, или сталкиваемся с null. Выглядит немного массивно, но создать запрос мощнее не вышло, приняли решение оставить.

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -66,44 +78,82 @@ public Collection<Film> getAllFilms() {
                 .collect(Collectors.toList());
     }
 
+    public List<Film> getTopPopularFilms(Integer count, Integer genreId, Integer year) {
```
В чём отличие от метода `getListPopularFilms`?

EvgeniyTomilov replied:
удален

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -66,44 +78,82 @@ public Collection<Film> getAllFilms() {
+            throw new ObjectNotFoundException("Режиссер у фильма не указан ");
+        }
+        switch (sortBy) {
+            case "year":
```
Здесь стоит перейти к перечислению

LipatovKir replied:
добавили

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,121 @@
+
+@Service("ReviewService")
+@Slf4j
+public class ReviewService {
```
Код-стайл в данном классе надо поправить

EvgeniyTomilov replied:
Код-стайл поправлен

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,121 @@
+            log.info("Отзыв " + id + " не найден");
+            throw new ResponseStatusException(HttpStatus.NOT_FOUND);
+        }
+        feedStorage.addEvent(getReviewById(id).getUserId(), EventTypes.REVIEW, Operations.REMOVE, id);
```
Рекомендую вынести результат `getReviewById(id).getUserId()` в отдельную переменную

AnnaTsepeleva replied:
поправила

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -6,18 +6,24 @@
 @Service
 @Slf4j
 public class UserService {
+    private Set<Long> friends = new HashSet<>();
```
Данное поле здесь не нужно, так как друзья хранятся в БД

EvgeniyTomilov replied:
поле удаленно

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -41,6 +41,8 @@ public void addFriend(Long friendId) {
     }
 
```
Поле `friends` более не нужно, так как друзья хранятся в БД

EvgeniyTomilov replied:
поле удаленно

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -34,9 +40,19 @@ public User updateUser(User user) {
     public void deleteUser(Long id) {
         if (contains(id)) {
             userStorage.delete(id);
+            getAllUsers()
```
Приведённый цикл выглядит сложно, воспользуйтесь каскадным удалением в БД.
Когда удалите пользователя - данный объект так же удалится и из всех связанных таблиц

Ekaterina-Kirpi replied:
Исправлено., добавлено каскадное удаление из БД

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -68,11 +85,12 @@ public void addFriend(Long id, Long friendId) {
             } else {
                 log.info("Пользователь " + friendId + " не найден");
-                throw new ResponseStatusException(HttpStatus.NOT_FOUND);
+                //throw new ResponseStatusException(HttpStatus.NOT_FOUND);
```
Неиспользуемый код надо убрать

Slavaniki replied:
исправлено

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/director/DirectorDbStorage.java:
```diff
@@ -0,0 +1,164 @@
+        this.jdbcTemplate = jdbcTemplate;
+    }
+
+    private Director rowMapToDirector(ResultSet resultSet, int i) throws SQLException {
```
Рекомендую вынести методы маппинга в отдельный класс, который бы отвечал за перевод объектов из `resultSet` в представление модели данных Вашего приложения

Slavaniki replied:
Мы командой решили так оставить

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/director/DirectorDbStorage.java:
```diff
@@ -0,0 +1,164 @@
+        String sep = ",";
+        String str = listID.stream().map(Object::toString)
+                .collect(Collectors.joining(sep));
+        String sqlQueryFilmDirectors = "select films_directors.film_id, directors.id, directors.name from films_directors, directors where films_directors.director_id = directors.id and films_directors.film_id in (" + str + ") order by films_directors.film_id";
```
Рекомендую воспользоваться оператором `group by` - за счёт группировки работать с результатом запроса станет проще

LipatovKir replied:
Мы командой решили так оставить

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/feed/FeedDbStorage.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    public Event mapRowFeed(ResultSet rs, int rowNum) throws SQLException {
+        LocalDate birthday;
```
Для чего здесь данный объект?

Slavaniki replied:
исправлено

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/feed/FeedDbStorage.java:
```diff
@@ -0,0 +1,87 @@
+
+@Slf4j
+@Component("feedDbStorage")
+@Primary
```
Данная аннотация так же не нужна, так как реализацию `inMemory` надо удалить

Slavaniki replied:
исправлено

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/feed/FeedDbStorage.java:
```diff
@@ -0,0 +1,87 @@
+public class FeedDbStorage implements FeedStorage {
+    private final JdbcTemplate jdbcTemplate;
+
+    @Autowired
```
Если хотите - можно не ставить `@Autowired`, spring поставит её за Вас

Slavaniki replied:
исправлено

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -97,7 +188,12 @@ public Film add(Film film) {
                     filmId,
                     genre.getId());
         });
```
Выполнять запросы в цикле в методе `add` неверно
Перепишите запрос таким образом, чтобы обновление всех объектов выполнялось за 1 запрос, для этого воспользуйтесь вставкой нескольких значений через `values` или через `batchUpdate`

AnnaTsepeleva replied:
сделала новый метод setGenresInDB в нем реализация через batchUpdate

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -125,22 +221,33 @@ public Optional<Film> update(Film film) {
+        if (film.getDirectors() != null) {
+            String updateDirectorQuery =
+                    "INSERT INTO films_directors (film_id, director_id) VALUES (?, ?)";
+            film.getDirectors().forEach(director -> {
```
Здесь аналогично, надо переписать под 1 запрос к БД

AnnaTsepeleva replied:
использовала метод setDIrectorsInDB в нем реализация через batchUpdate

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -125,22 +221,33 @@ public Optional<Film> update(Film film) {
+        if (film.getGenres() != null) {
+            String genreSqlQuery =
+                    "INSERT INTO genre (filmId, genreId) " + "VALUES (?, ?)";
+            film.getGenres().forEach(genre -> {
```
Здесь аналогично, надо переписать под 1 запрос к БД

AnnaTsepeleva replied:
сделала новый метод setGenresInDB в нем реализация через batchUpdate

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -39,15 +65,75 @@ private Film rowMapFilm(ResultSet rs) throws SQLException {
+                        .id(rs.getInt("ratingMPAId"))
+                        .name(rs.getString("ratingName"))
+                        .build())
+                .genres(getGenresOfFilm(filmId))
```
Вы вызываете метод `rowMapFilm` в разных циклах в коде, что приводит к выполнению метода `getGenresOfFilm` N раз. Надо переписать логику в коде, чтобы выполнялось не более 2-3 запросов к БД

KirPyt replied:
Замечание всё ещё не исправлено

Slavaniki replied:
Так как жанров может быть много, то в задании было предложено сделать отдельную таблицу для жанров, и так как жанры фильма могут добавляться/удаляться, необходимо делать запрос к этой таблице каждый раз

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -7,10 +7,7 @@
-import java.util.HashMap;
-import java.util.Optional;
+import java.util.*;
 
```
Реализацию `inMemory` удаляем, её поддержка более не нужна

EvgeniyTomilov replied:
Реализация удалена

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/likes/LikesDbStorage.java:
```diff
@@ -67,4 +69,14 @@ public Set<Long> getTopFilmLikes() {
\ No newline at end of file
+}
+
+
```
Надо убрать лишние переносы и лишний код

EvgeniyTomilov replied:
лишний и не используемый код, переносы  удалены

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/review/ReviewDbStorage.java:
```diff
@@ -0,0 +1,152 @@
+            return Optional.empty();
+        }
+        log.info("В базе данных найден отзыв: {}", reviews.get(0));
+        return Optional.of(reviews.get(0));
```
Не рекомендую оборачивать в `Optional` коллекции, т.к коллекции сами по себе уже являются "коробками" для данных

LipatovKir replied:
исправили

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/InMemoryUserStorage.java:
```diff
@@ -1,9 +1,10 @@
 package ru.yandex.practicum.filmorate.storage.user;
 
-import lombok.extern.slf4j.Slf4j;
+/*import lombok.extern.slf4j.Slf4j;
```
Хранилище `inMemory` удаляем

EvgeniyTomilov replied:
InMemoryUserStorage, удалён

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -179,4 +186,36 @@ public List<User> getCommonFriends(Long id, Long otherId) {
+    public Collection<Film> getRecommendations(Long id) {
+        Collection<Film> result = new HashSet<>();
+        Collection<Film> films = filmDbStorage.getAll();
+        getAll().forEach(user -> {
```
Вызывать запросы в цикле плохая идея, которая приводит к N запросам
Перепишите логику и запрос к БД, чтобы можно было выполнить всё не более чем за 2-3 запроса, для этого попробуйте работать не с 1 конкретным значением, а сразу со списками значений, тогда в запросах к БД можно будет воспользоваться оператором IN

KirPyt replied:
Замечание всё ещё не учтено

Slavaniki replied:
исправлено

KirPyt reviewed (commented):
Приветствую всех) Я Ваш ревьюер на данную работу.

Проект выполнен неплохо, но есть некоторые мысли, которые у меня сформировались при проверке Вашего проекта:

1. Стоит попробовать практику ревью кода друг друга, это очень полезный навык. В целом можно позадавать друг другу вопросы о том, почему здесь или там использовано то или иное решение и т.д. Так как это Ваша первая командная работа, то стоит попробовать оценить, насколько у Вас хорошо получилось наладить совместную разработку, разбиение реализации фич между собой и навык перекрёстного ревью.

2. Оцените оптимальность sql запросов, в том числе подвыборки данных там, где можно воспользоваться оператором join. Оцените лишние итерации в самом коде (например, поиск пересечений объектов в логике кода). Циклы с запросами к БД, которые порождают N запросов, такого быть не должно.

3. Обязательно проверьте общий код стайла Вашего проекта - код стайл во всем проекте должен быть выдержан в едином стиле, не содержать лишних переносов, закомментированных участков кода, пробелов и т.д

4. Не забывайте оставлять хорошие комментарии в коде, это поможет всем, и тем кто будет работать с Вашим кодом в дальнейшем, и Вам, когда и если будете писать документацию к своему коду

Так же некоторые замечания или советы актуальны и для остальных частей кода, поэтому рекомендую обратить внимание и на другие аналогичные места в коде, чтобы всё поправить.

Если у кого то из Вас возникнут вопросы, то пишите мне в Пачку. Можете связаться как Вам удобно, либо организованно через тимлида, либо задать вопросы отдельно.

EvgeniyTomilov committed:
sql -  большими буквами

EvgeniyTomilov committed:
Merge branch 'add-search' into develop

EvgeniyTomilov committed:
sql - большими буквами
InMemoryUserStorage - удален
лишние пробелы удалены
текст выровнен
удалены 2 неиспользуемых метода из FilmDbStorage

Slavaniki committed:
убраны лишние аннотации и закомменченное
удалён старый user storage
создан класс RowMapper, но не использованы его методы

Slavaniki committed:
checkstyle

EvgeniyTomilov committed:
лишние required = false удалены.
@Validated над классом добавлен

AnnaTsepeleva committed:
Commit by AS

EvgeniyTomilov committed:
чекстаил

EvgeniyTomilov committed:
fix

EvgeniyTomilov committed:
checkstyle

EvgeniyTomilov committed:
fix

EvgeniyTomilov committed:
checkstyle

EvgeniyTomilov committed:
checkstyle

Slavaniki committed:
добавление логирования

LipatovKir committed:
добавить enum

LipatovKir committed:
убрать коллекцию

LipatovKir committed:
убрать коллекцию

LipatovKir committed:
убрать коллекцию

LipatovKir committed:
убрать коллекцию

Ekaterina-Kirpi committed:
каскадное удаление пользователя и друзей

AnnaTsepeleva committed:
Commit by AS

EvgeniyTomilov committed:
checkStyle

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/RatingController.java:
```diff
@@ -20,13 +21,16 @@ public class RatingController {
     public MPA getMpaById(@PathVariable Integer id) {
         if (ratingService.getRatingById(id).isPresent()) {
+            log.info("Получение рейтинга с id: " + id);
             return ratingService.getRatingById(id).get();
```
Рекомендую вынести промежуточный результат `ratingService.getRatingById(id)` в переменную

LipatovKir replied:
исправлено

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -28,49 +32,62 @@ public User updateUser(@Valid @RequestBody User user) {
 
     @PutMapping("/{id}/friends/{userId}")
-    public void addFriend(@Valid @PathVariable Long id, @Valid @PathVariable Long userId) {
+    public void addFriend(@PathVariable Long id, @Valid @PathVariable Long userId) {
```
У второго параметра не убрали `@Valid`

LipatovKir replied:
исправлено

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -28,49 +32,62 @@ public User updateUser(@Valid @RequestBody User user) {
 
     @DeleteMapping("/{id}/friends/{userId}")
-    public void deleteFriend(@Valid @PathVariable Long id, @Valid @PathVariable Long userId) {
+    public void deleteFriend(@PathVariable Long id, @Valid @PathVariable Long userId) {
```
У второго параметра не убрали `@Valid`

LipatovKir replied:
исправлено

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -28,49 +32,62 @@ public User updateUser(@Valid @RequestBody User user) {
-
     @GetMapping("/{id}/friends/common/{userId}")
-    public Collection<User> getListSharedFriends(@Valid @PathVariable Long id, @Valid @PathVariable Long userId) {
+    public Collection<User> getListSharedFriends(@PathVariable Long id, @Valid @PathVariable Long userId) {
```
У второго параметра не убрали `@Valid`

LipatovKir replied:
исправлено

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/error/ErrorHandler.java:
```diff
@@ -31,9 +32,16 @@ public Map<String, String> handleNotFoundError(final NoSuchElementException e) {
+        return new ErrorResponse("404", e.getMessage());
+    }
+
     @ExceptionHandler({IllegalArgumentException.class, NullPointerException.class})
```
Когда Вы указываете в `@ExceptionHandler` конкретные перехватываемые исключения, то только они и будут попадать в этот хендлер. Вам нужно убрать указанные исключения и оставить только `@ExceptionHandler`, таким образом в этот хендлер будут лететь все исключения (проверяемые + непроверяемые), таким образом можно будет их корректно отработать

LipatovKir replied:
Мы решили удалить этот хэндлер, исключения отлавливаются остальными.

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/Director.java:
```diff
@@ -0,0 +1,23 @@
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+@Validated
```
`@Validated` здесь ну нужна

LipatovKir replied:
исправлено

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -3,17 +3,20 @@
 @Data
 @Builder
 @AllArgsConstructor
+@Validated
```
`@Validated` здесь ну нужна

LipatovKir replied:
исправлено

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/Review.java:
```diff
@@ -0,0 +1,27 @@
+@Data
+@Builder
+@AllArgsConstructor
+@Validated
```
`@Validated` здесь ну нужна

LipatovKir replied:
исправлено

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -3,17 +3,18 @@
 @Data
 @Builder
 @AllArgsConstructor
+@Validated
```
`@Validated` здесь ну нужна

LipatovKir replied:
исправлено

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,34 +1,40 @@
+    private static final String FILM_NOT_FOUND = "Фильм не найден № ";
 
-    public FilmService(LikesStorage likesStorage) {
+    public FilmService(FilmStorage filmStorage, UserStorage userStorage, LikesStorage likesStorage, DirectorService directorService, FeedStorage feedStorage) {
```
Если хотите, можно воспользоваться `@RequiredArgsConstructor`

LipatovKir replied:
исправлено

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/GenreService.java:
```diff
@@ -17,7 +17,6 @@ public class GenreService {
 
     @Autowired
-
     public GenreService(@Qualifier("GenresDbStorage") GenresStorage genresStorage) {
```
`@Qualifier("GenresDbStorage")` более не нужна

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,117 @@
+public class ReviewService {
+
+    @Autowired
+    @Qualifier(value = "filmDbStorage")
```
` @Qualifier` больше не нужны, так как других таких хранилищ нет

LipatovKir replied:
исправлено

KirPyt reviewed (commented):
Отлично, рад, что получилось учесть большую часть замечаний! Остались ещё некоторые замечания, которые надо поправить

LipatovKir committed:
исправить недочеты

LipatovKir committed:
исправить недочеты

LipatovKir committed:
исправить недочеты

Slavaniki committed:
исправление рекомендаций

Slavaniki committed:
добавлен комментарий про таблицу жанров

LipatovKir committed:
править хэндлер

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -22,32 +25,105 @@
                         .name(rs.getString("RATINGNAME"))
                         .build())
-                .genres(getGenresOfFilm(filmId))
+                .genres(getGenresOfFilm(filmId)) //Так как жанров может быть много, то в задании было предложено сделать
```
Что в итоге приводит к N запросам (в каждом запросе, который использует данный метод!) - это неоптимальное решение. Вы могли бы создать все необходимые объекты, а потом в конце просто запросить отдельно 1 запросом все жанры и затем в отдельном методе в цикле их добавить уже в программном коде. Тот комментарий, который Вы оставили позволяет сказать о том, что в процессе запросов жанров у Вас они будут каждый раз добавляться и удаляться, что в итоге после выполнения всех операций приведёт к неконсистентности данных, так как фильм, который был добавлен, например, в середине списка, и наполнен жанрами, может обладать другим списком жанров, по сравнению с последним фильмом в цикле.

EvgeniyTomilov replied:
Слава
Добавил новый класс FilmGenre, который содержит поля filmId и genre 
Убрал в мапперах заполнение жанров
Добавил метод получения списка FilmGenre и маппер для этого класса FilmGenre
И собственно основной метод, который заполняет фильмы жанрами, в нём с помощью этого списка FilmGenre мы собственно по очереди заполняем каждый фильм и получается, что запрос к базе был только один (на получение этого списка)
Ну и эти методы вызываем в нужных нам методах по типу getAll()

KirPyt reviewed (commented):
Описал важный момент по коду, который настоятельно рекомендую проработать. В остальном проект очень хорошо выполнен и замечания с некоторыми советами учтены.

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/error/ErrorHandler.java:
```diff
@@ -31,13 +32,11 @@ public Map<String, String> handleNotFoundError(final NoSuchElementException e) {
 
-    @ExceptionHandler({IllegalArgumentException.class, NullPointerException.class})
-    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
-    public Map<String, String> handleInternalError(final RuntimeException e) {
```
Вы полностью убрали 500-ый хендлер, вместо исправления данного замечания

EvgeniyTomilov replied:
Восстановили 500-ый хендлер.

EvgeniyTomilov committed:
fix

Slavaniki committed:
исправление многократных вызовов к таблице жанров

Slavaniki committed:
checkstyle

EvgeniyTomilov committed:
fix

EvgeniyTomilov committed:
Merge remote-tracking branch 'origin/develop' into develop

EvgeniyTomilov committed:
checkStyle

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/error/ErrorHandler.java:
```diff
@@ -38,5 +38,15 @@ public ErrorResponse handleObjectNotFoundException(final ObjectNotFoundException
         return new ErrorResponse("404", e.getMessage());
     }
+
+    @ExceptionHandler ({IllegalArgumentException.class, NullPointerException.class})
```
Из `@ExceptionHandler` надо убрать конкретные отлавливаемые исключения, так как таким образом Вы конкретизируете, какие ошибки надо отлавливать, а весь смысл `Throwable` в том, что мы ловим все ошибки, чтобы их корректно обрабатывать

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/FilmGenre.java:
```diff
@@ -0,0 +1,14 @@
+@Builder
+@Data
+@AllArgsConstructor
+public class FilmGenre {
```
В данной модели нет необходимости, так как у нас есть связь между таблицами жанров и фильмов, так что стоит просто сформировать корректный sql запрос

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -221,10 +227,26 @@ public void delete(Long id) {
-        String sqlQueryGetGenres = "SELECT GENREID FROM GENRE WHERE FILMID = ?";
-        return jdbcTemplate.queryForList(sqlQueryGetGenres, Integer.class, filmId)
-                .stream().map(genreStorage::getGenreById).collect(Collectors.toCollection(LinkedHashSet::new));
+    public List<FilmGenre> getAllGenresOfFilm() {
```
Стоит передавать только конкретный набор фильмов для отбора жанров, так как сейчас Вы получаете все жанры для всех фильмов, но может потребоваться организовать наполнение жанрами только некоторых фильмов, а не всех, и получается, что Вы всегда обрабатываете список из всех жанров. Для решения данной проблемы воспользуйтесь оператором IN в sql запросе

KirPyt reviewed (commented):
Указал на некоторые моменты по правкам, предлагаю попробовать разобраться в них
